keps/sig-network/3015-node-level-topology/README.md

```diff
@@ -0,0 +1,458 @@
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The `Local` policy for `service.spec.externalTrafficPolicy` and
+`service.spec.internalTrafficPolicy` allows configuring services so
+that they will always deliver to local endpoints, avoiding an
+unnecessary hop to another node when the request could just as easily
+be processed locally.
+
+However, they both fail entirely if there are no endpoints for the
+service on the node. This tradeoff is unacceptable for some use cases.
+
+The [Topology Aware Hints feature] provides a way of signaling to
+kube-proxy that it should optimize service delivery on a
+per-availability zone basis, but does not provide any way to optimize
+on any other basis.
+
+This KEP adds a new topology hint, to tell kube-proxy that a Service
+is expected to have an endpoint on every node most of the time, and so
+it should optimize traffic by always routing to a local endpoint _when
+one is available_, but using other topology hints or ordinary
+`Cluster` traffic policy when not.
+
+[Topology Aware Hints feature]: ../2433-topology-aware-hints/
+
+## Motivation
+
+### Goals
+
+- Allow configuring a service so that connections will be delivered to
+  a local endpoint when possible, and a remote endpoint if not.
+
+```
+<<[UNRESOLVED deprecate-iTP]>>
+
+- Deprecate `internalTrafficPolicy`? It's clear that the DNS use case
+  given in the Internal Traffic Policy KEP is not actually a good use
+  case for Internal Traffic Policy, because no one wants the behavior
+  of "I'd rather have DNS requests get dropped than have them go to
+  another node". But without the DNS use case, it's not clear that
+  there's really a strong argument for Internal Traffic Policy at all.
+
+<<[/UNRESOLVED]>>
+```
+
+### Non-Goals
+
+- Providing any improvements over `externalTrafficPolicy: Local` for
+  LoadBalancer services. Consensus is that [Proxy Terminating
+  Endpoints] should solve the problems that made
+  `externalTrafficPolicy: Local` unreliable for some cases.
```

Comment:
Would the same apply for `internalTrafficPolicy: Local`?

Reply:
No. (It's important to remember that iTP:Local and eTP:Local are really totally different things; they have the same name because the very-low-level kube-proxy implementation details are the same, but at a high level they are used in completely different ways.)

The problem with eTP:Local is a race condition where the cloud LB is technically violating the contract of the eTP:Local service by sending traffic to a node that reports that it has no live endpoints. We can't easily fix the LB to not have that race condition, so PTE fixes the problem by making endpoints continue working even after the node claims to have no endpoints.

The problem with iTP:Local is not a race condition, and doesn't involve anyone doing anything illegal; the problem is that even with PTE (or `terminationGracePeriodSeconds: 0`), there is still a gap between when the old endpoint exits and the new one starts serving, and the iTP:Local service is not available during that gap.